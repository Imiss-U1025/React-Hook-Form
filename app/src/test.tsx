import * as React from 'react';
import { useFieldArray, useForm } from 'react-hook-form';
import { Control } from '../../src/types';

let outputData: object = {};
let watchedData: object[] = [];

const Component = () => {
  const { register, handleSubmit, watch } = useForm<{
    test?: string;
    test1?: string;
    test2?: string;
    test3?: string;
    test4: string;
  }>();

  watchedData.push(watch());

  return (
    <form
      onSubmit={handleSubmit((data) => {
        console.log(data)
        outputData = data;
      })}
    >
      <input {...register('test')} disabled />
      <input
        disabled={true}
        value={'test'}
        type={'checkbox'}
        {...register('test1')}
      />
      <input
        disabled={true}
        value={'test'}
        type={'radio'}
        {...register('test2')}
      />
      <select {...register('test3')} disabled />
      <input {...register('test4')} data-testid={'input'} />
      <button>Submit</button>
    </form>
  );
};

export default Component;
