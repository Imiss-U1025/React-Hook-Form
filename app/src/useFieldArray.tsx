import React from 'react';
import { useForm, useFieldArray, Controller } from 'react-hook-form';

let renderCounter = 0;

const UseFieldArray: React.FC = (props: any) => {
  const { control, handleSubmit } = useForm<{
    data: object[];
  }>({
    ...(props.match.params.mode === 'default'
      ? {
          defaultValues: {
            data: [{ name: 'test' }, { name: 'test1' }, { name: 'test2' }],
          },
        }
      : {}),
  });
  const { fields, append, prepend, swap, move, insert, remove } = useFieldArray(
    {
      control,
      name: 'data',
    },
  );
  const onSubmit = () => {};

  renderCounter++;

  console.log(fields);

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {fields.map((data, index) => (
        <div key={data.id}>
          {index % 2 ? (
            <input
              className={`input${index}`}
              key={data.id}
              name={`data[${index}].name`}
              defaultValue={data.id}
              data-id={renderCounter}
            />
          ) : (
            <Controller
              as={<input />}
              control={control}
              name={`data[${index}].name`}
              defaultValue={data.id}
              data-id={renderCounter}
            />
          )}
          <button className={`delete${index}`}>delete</button>
        </div>
      ))}

      <button id="append" onClick={() => append({ test: 1 })}>
        append
      </button>

      <button id="prepend" onClick={() => prepend({ test: 1 })}>
        prepend
      </button>

      <button id="swap" onClick={() => swap(1, 3)}>
        swap
      </button>

      <button id="move" onClick={() => move(1, 3)}>
        move
      </button>

      <button id="insert" onClick={() => insert(2, { test: 1 })}>
        insert
      </button>

      <button id="remove" onClick={() => remove(1)}>
        remove
      </button>

      <button id="remove" onClick={() => remove()}>
        remove all
      </button>

      <div id="renderCounter">{renderCounter}</div>
    </form>
  );
};

export default UseFieldArray;
