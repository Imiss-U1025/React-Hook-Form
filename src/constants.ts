import { ValidationMode } from './types/form';

export const VALIDATION_MODE: ValidationMode = {
  onBlur: 'onBlur',
  onChange: 'onChange',
  onSubmit: 'onSubmit',
  all: 'all',
};

export const VALUE = 'value';

export const UNDEFINED = 'undefined';

export const EVENTS = {
  BLUR: 'blur',
  CHANGE: 'change',
  INPUT: 'input',
};

export const SELECT = 'select';

export const INPUT_VALIDATION_RULES = {
  max: 'max',
  min: 'min',
  maxLength: 'maxLength',
  minLength: 'minLength',
  pattern: 'pattern',
  required: 'required',
  validate: 'validate',
};
