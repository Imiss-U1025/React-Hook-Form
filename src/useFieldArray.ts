import * as React from 'react';
import { useFormContext } from './useFormContext';
import { isMatchFieldArrayName } from './logic/isNameInFieldArray';
import { appendId, mapIds } from './logic/mapIds';
import getIsFieldsDifferent from './logic/getIsFieldsDifferent';
import transformToNestObject from './logic/transformToNestObject';
import getFieldsValues from './logic/getFieldsValues';
import get from './utils/get';
import isUndefined from './utils/isUndefined';
import removeArrayAt from './utils/remove';
import moveArrayAt from './utils/move';
import swapArrayAt from './utils/swap';
import prependAt from './utils/prepend';
import insertAt from './utils/insert';
import { FieldValues, Control, UseFieldArrayProps, WithFieldId } from './types';

export function useFieldArray<
  FormArrayValues extends FieldValues = FieldValues,
  ControlProp extends Control = Control
>({ control, name }: UseFieldArrayProps<ControlProp>) {
  const methods = useFormContext();
  const {
    resetFieldArrayFunctionRef,
    fieldArrayNamesRef,
    fieldsRef,
    defaultValuesRef,
    removeRef,
    errorsRef,
    isDirtyRef,
    touchedFieldsRef,
    readFormStateRef,
  } = control || methods.control;
  const memoizedDefaultValues = React.useRef(
    get(defaultValuesRef.current, name, []),
  );
  const [fields, setField] = React.useState<
    WithFieldId<Partial<FormArrayValues>>[]
  >(mapIds(memoizedDefaultValues.current));
  const getFieldValuesByName = <T, K extends keyof T>(fields: T, name: K) =>
    transformToNestObject(getFieldsValues(fields))[name];

  const resetFields = (
    flagOrFields?: WithFieldId<Partial<FormArrayValues>>[],
  ) => {
    if (readFormStateRef.current.dirty) {
      isDirtyRef.current = isUndefined(flagOrFields)
        ? true
        : getIsFieldsDifferent(flagOrFields, memoizedDefaultValues.current);
    }

    for (const key in fieldsRef.current) {
      if (isMatchFieldArrayName(key, name)) {
        if (
          fieldsRef.current[key] &&
          (fieldsRef.current[key] as any).ref instanceof HTMLElement
        ) {
          removeRef(fieldsRef.current[key], true);
          delete fieldsRef.current[name];
        }
      }
    }
  };

  const append = (value: WithFieldId<Partial<FormArrayValues>>) => {
    if (readFormStateRef.current.dirty) {
      isDirtyRef.current = true;
    }
    setField([...fields, appendId(value)]);
  };

  const prepend = (value: WithFieldId<Partial<FormArrayValues>>) => {
    resetFields();
    setField(prependAt(fields, appendId(value)));
    if (errorsRef.current[name]) {
      errorsRef.current[name] = prependAt(errorsRef.current[name]);
    }
    if (readFormStateRef.current.touched && touchedFieldsRef.current[name]) {
      touchedFieldsRef.current[name] = prependAt(
        touchedFieldsRef.current[name],
      );
    }
  };

  const remove = (index?: number) => {
    resetFields(
      removeArrayAt(getFieldValuesByName(fieldsRef.current, name), index),
    );
    setField(removeArrayAt(fields, index));
    if (errorsRef.current[name]) {
      errorsRef.current[name] = removeArrayAt(errorsRef.current[name], index);
    }
    if (readFormStateRef.current.touched && touchedFieldsRef.current[name]) {
      touchedFieldsRef.current[name] = removeArrayAt(
        touchedFieldsRef.current[name],
        index,
      );
    }
  };

  const insert = (
    index: number,
    value: WithFieldId<Partial<FormArrayValues>>,
  ) => {
    resetFields();
    setField(insertAt(fields, index, appendId(value)));
    if (errorsRef.current[name]) {
      errorsRef.current[name] = insertAt(errorsRef.current[name], index);
    }
    if (readFormStateRef.current.touched && touchedFieldsRef.current[name]) {
      touchedFieldsRef.current[name] = insertAt(
        touchedFieldsRef.current[name],
        index,
      );
    }
  };

  const swap = (indexA: number, indexB: number) => {
    const fieldValues = getFieldValuesByName(fieldsRef.current, name);
    swapArrayAt(fieldValues, indexA, indexB);
    resetFields(fieldValues);
    swapArrayAt(fields, indexA, indexB);
    setField([...fields]);
    if (errorsRef.current[name]) {
      swapArrayAt(errorsRef.current[name], indexA, indexB);
    }
    if (readFormStateRef.current.touched && touchedFieldsRef.current[name]) {
      swapArrayAt(touchedFieldsRef.current[name], indexA, indexB);
    }
  };

  const move = (from: number, to: number) => {
    const fieldValues = getFieldValuesByName(fieldsRef.current, name);
    moveArrayAt(getFieldValuesByName(fieldsRef.current, name), from, to);
    resetFields(fieldValues);
    moveArrayAt(fields, from, to);
    setField([...fields]);
    if (errorsRef.current[name]) {
      moveArrayAt(errorsRef.current[name] as [], from, to);
    }
    if (readFormStateRef.current.touched && touchedFieldsRef.current[name]) {
      moveArrayAt(touchedFieldsRef.current[name], from, to);
    }
  };

  const reset = () => {
    resetFields();
    memoizedDefaultValues.current = get(defaultValuesRef.current, name, []);
    setField(mapIds(memoizedDefaultValues.current));
  };

  React.useEffect(() => {
    const resetFunctions = resetFieldArrayFunctionRef.current;
    const fieldArrayNames = fieldArrayNamesRef.current;
    fieldArrayNames.add(name);
    resetFunctions[name] = reset;

    return () => {
      resetFields();
      delete resetFunctions[name];
      fieldArrayNames.delete(name);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [name]);

  return {
    swap,
    move,
    prepend,
    append,
    remove,
    insert,
    fields,
  };
}
