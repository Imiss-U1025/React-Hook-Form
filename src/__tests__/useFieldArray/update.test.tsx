import * as React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { act, renderHook } from '@testing-library/react-hooks';

import { VALIDATION_MODE } from '../../constants';
import { useFieldArray } from '../../useFieldArray';
import { useForm } from '../../useForm';
import { mockGenerateId } from '../useFieldArray.test';

describe('update', () => {
  beforeEach(() => {
    mockGenerateId();
  });

  it('should update dirtyFields fields correctly', async () => {
    let dirtyInputs = {};
    const Component = () => {
      const {
        register,
        control,
        formState: { dirtyFields },
      } = useForm<{
        test: { value: string }[];
      }>({
        defaultValues: {
          test: [
            { value: 'plz change' },
            { value: 'dont change' },
            { value: 'dont change' },
          ],
        },
      });
      const { fields, update } = useFieldArray({
        control,
        name: 'test',
      });

      dirtyInputs = dirtyFields;

      return (
        <form>
          {fields.map((field, i) => (
            <input key={field.id} {...register(`test.${i}.value` as const)} />
          ))}
          <button type="button" onClick={() => update(0, { value: 'changed' })}>
            update
          </button>
          {dirtyFields.test?.length && 'dirty'}
        </form>
      );
    };

    render(<Component />);

    fireEvent.click(screen.getByRole('button'));

    await waitFor(() => screen.getByText('dirty'));

    expect(dirtyInputs).toEqual({
      test: [{ value: true }],
    });
  });

  it.each(['isDirty', 'dirtyFields'])(
    'should be dirtyFields when value is appended with %s',
    () => {
      let isDirtyValue;
      let dirtyValue;

      const Component = () => {
        const {
          register,
          control,
          formState: { isDirty, dirtyFields },
        } = useForm<{
          test: { test: string }[];
        }>({
          defaultValues: {},
        });
        const { fields, update } = useFieldArray({
          control,
          name: 'test',
        });

        isDirtyValue = isDirty;
        dirtyValue = dirtyFields;

        return (
          <form>
            {fields.map((field, index) => {
              return (
                <input
                  key={field.id}
                  {...register(`test.${index}.test` as const)}
                />
              );
            })}
            <button
              type={'button'}
              onClick={() => update(2, { test: 'test1' })}
            >
              update
            </button>
          </form>
        );
      };

      render(<Component />);

      act(() => {
        fireEvent.click(screen.getByRole('button', { name: 'update' }));
      });

      expect(isDirtyValue).toBeTruthy();
      expect(dirtyValue).toEqual({
        test: [undefined, undefined, { test: true }],
      });
    },
  );

  it('should trigger reRender when user update input and is watching the all field array', () => {
    const watched: any[] = [];
    const Component = () => {
      const { register, watch, control } = useForm<{
        test: { value: string }[];
      }>();
      const { fields, update } = useFieldArray({
        control,
        name: 'test',
      });
      watched.push(watch());

      return (
        <form>
          {fields.map((field, i) => (
            <input key={field.id} {...register(`test.${i}.value` as const)} />
          ))}
          <button type="button" onClick={() => update(0, { value: '' })}>
            update
          </button>
        </form>
      );
    };

    render(<Component />);

    fireEvent.click(screen.getByRole('button', { name: /update/i }));

    expect(watched).toMatchSnapshot();
  });

  it('should return watched value with update and watch API', async () => {
    const renderedItems: any = [];
    const Component = () => {
      const { watch, register, control } = useForm<{
        test: { value: string }[];
      }>();
      const { fields, update } = useFieldArray({
        name: 'test',
        control,
      });
      const watched = watch('test');
      renderedItems.push(watched);
      return (
        <div>
          {fields.map((field, i) => (
            <div key={field.id}>
              <input {...register(`test.${i}.value` as const)} />
            </div>
          ))}
          <button onClick={() => update(0, { value: 'test' })}>update</button>
        </div>
      );
    };

    render(<Component />);

    fireEvent.click(screen.getByRole('button', { name: /update/i }));

    await waitFor(() =>
      expect(renderedItems).toEqual([
        undefined,
        [],
        [{ value: 'test' }],
        [{ value: 'test' }],
      ]),
    );
  });

  describe('with resolver', () => {
    it('should invoke resolver when formState.isValid true', async () => {
      const resolver = jest.fn().mockReturnValue({});

      const { result } = renderHook(() => {
        const { formState, control } = useForm({
          mode: VALIDATION_MODE.onChange,
          resolver,
        });
        const { update } = useFieldArray({ control, name: 'test' });
        return { formState, update };
      });

      result.current.formState.isValid;

      await act(async () => {
        result.current.update(0, { value: '1' });
      });

      expect(resolver).toBeCalledWith(
        {
          test: [{ value: '1' }],
        },
        undefined,
        {
          criteriaMode: undefined,
          fields: {
            test: [
              {
                value: {
                  mount: true,
                  name: 'test.0.value',
                  ref: {
                    name: 'test.0.value',
                    value: '1',
                  },
                  value: '1',
                },
              },
            ],
          },
          names: ['test.0.value'],
        },
      );
    });

    it('should not invoke resolver when formState.isValid false', () => {
      const resolver = jest.fn().mockReturnValue({});

      const { result } = renderHook(() => {
        const { formState, control } = useForm({
          mode: VALIDATION_MODE.onChange,
          resolver,
        });
        const { update } = useFieldArray({ control, name: 'test' });
        return { formState, update };
      });

      act(() => {
        result.current.update(0, { value: '1' });
      });

      expect(resolver).not.toBeCalled();
    });
  });
});
