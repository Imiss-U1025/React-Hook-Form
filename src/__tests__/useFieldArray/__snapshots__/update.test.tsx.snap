// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update should trigger reRender when user update input and is watching the all field array 1`] = `
Array [
  Object {},
  Object {
    "test": Array [],
  },
  Object {
    "test": Array [
      Object {
        "value": "",
      },
    ],
  },
  Object {
    "test": Array [
      Object {
        "value": "",
      },
    ],
  },
]
`;

exports[`update should update field array with multiple values 1`] = `
Array [
  Object {
    "firstName": "test1",
    "lastName": "test2",
  },
  Object {
    "firstName": "test3",
    "lastName": "test4",
  },
]
`;

exports[`update should update group input correctly 1`] = `
Array [
  Array [
    Object {
      "key": "0",
      "value": Object {
        "value": Object {
          "firstName": "bill",
          "lastName": "luo",
        },
      },
    },
  ],
  Array [
    Object {
      "key": "1",
      "value": Object {
        "value": Object {
          "firstName": "firstName",
          "lastName": "lastName",
        },
      },
    },
  ],
]
`;
